### PR TITLE
Fix ArrayIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by InvisibleProgrammer**
It is possible to put `VectorizedRowBatch` into illegal state by using field access. 

This PR fixes two places that can cause ArrayIndexOutOfBoundsException in case of the selected array is set but the size doesn't updated properly. 

### What changes were proposed in this pull request?
Fixed ArrayIndexOutOfBoundsException in `VectorTopNKeyOperator` and `VectorFilterOperator`.

### Why are the changes needed?
To avoid ArrayIndexOutOfBoundsException.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Added new tests. The tests failed with exception without the fix. 
- End to end test on the dev environment

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * use them properly.
+ * </p>
+ * Ticket to fix the issue: HIVE-26993
+ */
+public class TestVectorOperationProcess {
+
+  HiveConf hiveConf = new HiveConf();
+
+  @Test
+  public void testVectorFilterHasSelectedSmallerThanBatch() throws HiveException {
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:68]: why not do this in TestVectorFilterOperator?

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    FakeDataReader fdr = new FakeDataReader(1024, 3);
+    VectorizedRowBatch vrg = fdr.getNext();
+
+    vrg.selected = new int[] { 1, 2, 3, 4};
+
+    vfo.process(vrg, 0);
+  }
+
+  @Test
+  public void testTopNHasSelectedSmallerThanBatch() throws HiveException {
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:99]: There should be a test for a TopNKey as well. Test name is not declarative.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java [hunk 3]
```diff
@@ -4397,9 +4397,9 @@ public static Operator<? extends OperatorDesc> vectorizeFilterOperator(
         vContext, vectorFilterDesc);
   }
 
-  private static Operator<? extends OperatorDesc> vectorizeTopNKeyOperator(
-      Operator<? extends OperatorDesc> topNKeyOperator, VectorizationContext vContext,
-      VectorTopNKeyDesc vectorTopNKeyDesc) throws HiveException {
+  public static Operator<? extends OperatorDesc> vectorizeTopNKeyOperator(
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:4400]: why is this change?

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    VectorExpression ve3 = new FilterExprAndExpr();
+    ve3.setChildExpressions(new VectorExpression[] {ve1, ve2});
+    vfo.setFilterCondition(ve3);
+
+    FakeDataReader fdr = new FakeDataReader(1024, 3);
+    VectorizedRowBatch vrg = fdr.getNext();
+
+    vrg.selected = new int[] { 1, 2, 3, 4};
+
+    vfo.process(vrg, 0);
```
Comment (deniskuzZ) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: what does it test? should we have any asserts?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java (hunk 3 above)
Comment (InvisibleProgrammer) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:4400]: To be able to test the class. I used the same pattern as `VectorFilterOperator` has. You can see the example in `TestVectorFilterOperator.getAVectorFilterOperator`.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 1 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:68]: Good question. TestVectorFilterOperator has a comment that says it is only for fundamental logic and performance testing for VectorFilterOperator. Note: as I can see, I made a mistake and moved that test out of this class at a refactor tests. I'm going to move the comment to its proper place. 

The other reason is actually the root cause of the problem: the origin of the problem is that operators allow direct field access instead of using methods. So that, it is easily to put the object into invalid state. It is enough to update the selected list but do not update the size field. 

Now that fix fixed that issue. But currently, there is no guarantee that we won't create an other operator usage 3 months from now that forgets to update the size method and we will be in the exact same situation. 
I want to keep that fix as a temporary one and refactor the original code later in a way that doesn't allow field access and so that, I wanted to keep the related temporary fix tests in a single class. So that, when the issue is fixed, we can just simply delete that test class.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: It is a tricky one. The issue is that without our fix, it should throw an exception. Actually, I didn't find anything that we could assert on. Maybe if we change the fake data generator to not to use random data, we can modify the test to assert if the filtering operator properly filters out the records from the row batch. 
If you check the process method of the VectorizedFilterOperator, it saves it's own state at the beginning of the process method, process the row batch and restores it's original state. So, after the process, there is no difference in the state of the class that we test.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 2 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:99]: Do you have any suggestions?

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    VectorExpression ve2 = new FilterLongColEqualDoubleScalar(2, 0);
+    VectorExpression ve3 = new FilterExprAndExpr();
+    ve3.setChildExpressions(new VectorExpression[] {ve1, ve2});
+
+    FakeDataReader fdr = new FakeDataReader(1024, 3);
+    VectorizedRowBatch vrg = fdr.getNext();
+
+    vrg.selected = new int[] { 1, 2, 3, 4};
+
+    vfo.process(vrg, 0);
```
[the remark this answers is not in the document]
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:124]: And the same answer as above.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:99]: move it to the TestTopNKeyFilter test

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: maybe at least check that no exceptions are being thrown:
````
Assertions.assertDoesNotThrow(() -> {....})
````

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java (hunk 3 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/Vectorizer.java:4400]: can we mark it as @VisibleForTesting

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:68]: I think we still need the tests that cover the edge cases and would be convenient if they are not scattered all over the place.

Comment (kasakrisz) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:68]: I think a common base class can be introduced to extract the common parts of `TestVectorFilterOperator` and `TestVectorTNKOperator`

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 2 above)
Comment (kasakrisz) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:99]: `TestTopNKeyFilter` tests `TopNKeyFilter` only which is wrapped by `TopNKey` operator. It provides the logic of caching the top n values and TNK decides whether to forward a row based on `TopNKeyFilter.canForward` method.

I think it is worth having these tests in a class like `TestVectorTopNKeyOperator` if the goal is to test whether selected array is filled properly.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (kasakrisz) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: Using random data in test is a bad practice. It makes harder to reproduce failures and to find out root cause.
If we really insists adding these tests please use a constant data set and add assertions to the test cases.

Example: IIUC this test case test if the number of selected rows are less than the batch size. In this case an assert like
```
Assert.assertTrue(vrg.selected.length < vrg.size);
```
makes sense.

Please share your thoughts.

**Comment by kasakrisz** [comment]
Maybe I'm missing something but I haven't found any repro steps of the issue. Please share if there is any and create a q test from it. It can be also useful to use it instead of adding temporary unit tests.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: > Assert.assertTrue(vrg.selected.length < vrg.size);

Both VectorFilterOperator and VectorTopNKeyOperator restores their original state at the end of the process method. So I cannot do assert on that. 
I've tried multiple methods but couldn't figure out a great thing to assert on - yet. 

> Assertions.assertDoesNotThrow(() -> {....})

There is no such a thing in Java4. There is a possibility to annotate the test like this `@Test(expected = None.class)`  but IntelliJ offers me to remove unnecessary code. 

@kasakrisz , @deniskuzZ : what do you think about that?

Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: check this https://github.com/apache/hive/blob/6e25a0151d54d2119a2a71e1212f606c8ca55a37/iceberg/iceberg-catalog/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java#L124

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:68]: 👍 like Abstract/TestVectorOperator, that could include the FakeDataReader and other stuff

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: can't you add the suggested by Kristian check before calling the process?

Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: also can't you initialize VectorizedRowBatch by yourself, looks like FakeDataReader was created for some other reasons (simulate actual read ??? there is some sleep functionality)

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 2 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:99]: 👍

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 4 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java:95]: > can't you add the suggested by Kristian check before calling the process?

> Example: IIUC this test case test if the number of selected rows are less than the batch size. In this case an assert like

> Assert.assertTrue(vrg.selected.length < vrg.size);
> makes sense.

Our test case is about manipulating selected array without modifying the size field leads to an inconsistent state so the process method fails with exception. When I modify the selected array and leaves the size field, it is obvious that the selected.length will be smallar than the size.
I give a second chance to Assertions.assertDoesNotThrow

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 1 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java]: Introduced a new base class, `TestVectorOperator`. It contains FakeDataReader and HiveConf. I didn't found other parts of the code that worth extracting here as there are minor differences between those tests.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java (hunk 2 above)
Comment (InvisibleProgrammer) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/vector/TestVectorOperationProcess.java]: Introduced `TestVectorTopNKeyOperator`. 

I couldn't test if the selected array is filled properly as our issue is about the case when it is not filled properly.

**Comment by InvisibleProgrammer** [comment]
@deniskuzZ , @kasakrisz : 

I was able to reproduce it on a dev cluster and validated the change with manual testing.

Selects for the operator changes:
```sql
select count(*) from tbl_import where L_DISCOUNT = 0.0;

SELECT l_discount FROM tbl_import GROUP BY l_discount ORDER BY l_discount LIMIT 3;
```

Preparation steps: 
```sql
drop table if exists lineitem_text;

create external table lineitem_text
(L_ORDERKEY BIGINT,
 L_PARTKEY BIGINT,
 L_SUPPKEY BIGINT,
 L_LINENUMBER INT,
 L_QUANTITY DOUBLE,
 L_EXTENDEDPRICE DOUBLE,
 L_DISCOUNT DOUBLE,
 L_TAX DOUBLE,
 L_RETURNFLAG STRING,
 L_LINESTATUS STRING,
 L_SHIPDATE STRING,
 L_COMMITDATE STRING,
 L_RECEIPTDATE STRING,
 L_SHIPINSTRUCT STRING,
 L_SHIPMODE STRING,
 L_COMMENT STRING)
ROW FORMAT DELIMITED FIELDS TERMINATED BY '|' STORED AS TEXTFILE
LOCATION 's3a:/wonttell;

drop table if exists lineitem_acid;

create table lineitem_acid
(L_ORDERKEY BIGINT,
 L_PARTKEY BIGINT,
 L_SUPPKEY BIGINT,
 L_LINENUMBER INT,
 L_QUANTITY DOUBLE,
 L_EXTENDEDPRICE DOUBLE,
 L_DISCOUNT DOUBLE,
 L_TAX DOUBLE,
 L_RETURNFLAG STRING,
 L_LINESTATUS STRING,
 L_SHIPDATE STRING,
 L_COMMITDATE STRING,
 L_RECEIPTDATE STRING,
 L_SHIPINSTRUCT STRING,
 L_SHIPMODE STRING,
 L_COMMENT STRING)
CLUSTERED BY (L_ORDERKEY) into 10 buckets
STORED AS ORC TBLPROPERTIES('transactional'='true');

INSERT INTO TABLE lineitem_acid
SELECT * FROM lineitem_text;


create external table lineitem_copy
(L_ORDERKEY BIGINT,
 L_PARTKEY BIGINT,
 L_SUPPKEY BIGINT,
 L_LINENUMBER INT,
 L_QUANTITY DOUBLE,
 L_EXTENDEDPRICE DOUBLE,
 L_DISCOUNT DOUBLE,
 L_TAX DOUBLE,
 L_RETURNFLAG STRING,
 L_LINESTATUS STRING,
 L_SHIPDATE STRING,
 L_COMMITDATE STRING,
 L_RECEIPTDATE STRING,
 L_SHIPINSTRUCT STRING,
 L_SHIPMODE STRING,
 L_COMMENT STRING);

INSERT INTO TABLE lineitem_copy
SELECT * FROM lineitem_text;

ALTER TABLE lineitem_copy
REPLACE COLUMNS (L_ORDERKEY BIGINT, L_PARTKEY BIGINT, L_SUPPKEY BIGINT, L_LINENUMBER INT, L_QUANTITY DOUBLE, L_EXTENDEDPRICE DOUBLE, L_DISCOUNT DOUBLE, L_TAX DOUBLE);

create table lineitem_acid_copy
(L_ORDERKEY BIGINT,
 L_PARTKEY BIGINT,
 L_SUPPKEY BIGINT,
 L_LINENUMBER INT,
 L_QUANTITY DOUBLE,
 L_EXTENDEDPRICE DOUBLE,
 L_DISCOUNT DOUBLE,
 L_TAX DOUBLE)
 CLUSTERED BY (L_ORDERKEY) into 10 buckets
 STORED AS ORC TBLPROPERTIES('transactional'='true');

INSERT INTO TABLE lineitem_acid_copy
SELECT * FROM lineitem_copy;
 


drop table if exists tbl_export;

create table tbl_export (L_ORDERKEY BIGINT, L_PARTKEY BIGINT, L_SUPPKEY BIGINT, L_LINENUMBER INT, L_QUANTITY DOUBLE, L_EXTENDEDPRICE DOUBLE, L_DISCOUNT DOUBLE) partitioned by (L_TAX DOUBLE) clustered by (L_ORDERKEY) into 10 buckets;

insert into table tbl_export partition (L_TAX) select * from lineitem_acid_copy;


drop table if exists tbl_import;


create table tbl_import (L_ORDERKEY BIGINT, L_PARTKEY BIGINT, L_SUPPKEY BIGINT, L_LINENUMBER INT, L_QUANTITY DOUBLE, L_EXTENDEDPRICE DOUBLE, L_DISCOUNT DOUBLE) partitioned by (L_TAX DOUBLE) clustered by (L_ORDERKEY) into 10 buckets;


EXPORT TABLE tbl_export to '!!EXPORT_DIR!!/tmp/exportDir';

IMPORT TABLE tbl_import FROM '!!EXPORT_DIR!!/tmp/exportDir';

analyze table tbl_import compute statistics;

select count(*) from tbl_import;

delete from tbl_import where L_TAX = 0.0 and L_DISCOUNT = 0.0;
```

**Comment by InvisibleProgrammer** [comment]
@deniskuzZ , @kasakrisz : finally, it became all green. Can we move forward with it?

**Review by deniskuzZ — APPROVED** [review]
LGTM +1

**Review by kasakrisz — APPROVED** [review]
Thanks for confirming the patch @InvisibleProgrammer
